# import the necessary libary

In [17]:
import pandas as pd 
import re
import time
from sqlalchemy import create_engine
import requests
from bs4 import BeautifulSoup
import cloudscraper
import selenium.webdriver as webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import  By

create a connection to the website

In [18]:
url ='https://www.aliexpress.com/w/wholesale-laptop-.html?g=y&SearchText=laptop+&shpf_co=CN'
scarper = cloudscraper.create_scraper()
respond = scarper.get(url)
print(respond)

<Response [200]>


create a chrome driver connection to the website

In [40]:
chrome_option=Options()
drive_path=r"C:\Users\Merci\webscipping aliexpress\.ipynb_checkpoints\chromedriver-win64\chromedriver-win64\chromedriver.exe"
chrome_service= Service(drive_path)
driver = webdriver.Chrome(service=chrome_service,options=chrome_option)
page_source=driver.page_source
time.sleep(20)

In [27]:
product_name=[]
price=[]
item_rating=[]
previous_price=[]
store_link=[]
discount=[]
shipping_status=[]
totalsold=[]
rating=[]
totalpage =10
for page in range(totalpage+1):
    url=f"https://www.aliexpress.com/w/wholesale-phone.html?page={page}&g=y&SearchText=phone"
    driver.get(url)
    soup = BeautifulSoup(driver.page_source,'html.parser')
    totalphone=soup.find_all("div",class_="list--galleryWrapper--29HRJT4V")
    for phone in totalphone:
        product_name=phone.find("div", class_="multi--title--G7dOCj3").text.strip() if phone.find("div", class_="multi--title--G7dOCj3") else ''
        price=phone.find("div",class_="multi--price-sale--U-S0jtj").text.strip().replace('HUF','').replace(',','').replace('.','') if phone.find("div",class_="multi--price-sale--U-S0jtj") else ''
        totalsold=phone.find('span',class_="multi--trade--Ktbl2jB").text.strip() if phone.find('span',class_="multi--trade--Ktbl2jB").text.strip() else ''
        previous_price = phone.find('div',class_='multi--price-original--1zEQqOK').text.strip().replace('HUF','').replace(',','').replace('.','') if phone.find('div',class_='multi--price-original--1zEQqOK') else ''
        shipping_status= phone.find("div",class_="multi--serviceContainer--3vRdzWN").text.strip()if phone.find("div",class_="multi--serviceContainer--3vRdzWN") else 'No shipping'
        
        try:
            rating=phone.find('div',class_='multi--starList--t9_CAj2').text.strip() 
        except AttributeError:
            rating =''
        try:
            store_kink=phone.find("a",class_="cards--storeLink--XkKUQFS")['href']
        except(TypeError,KeyError):
            store_link=''
        try:
            store_name=phone.find("a",class_="cards--storeLink--XkKUQFS").text.strip()
        except AttributeError:
            store_name=''
        try:
            discount = phone.find("span",class_="multi--discount--3hksz5G").text.strip()
        except AttributeError:
            discount=''
        product_name= product_name.append( product_name)
        price=price.append(price)
        item_rating=item_rating.append(item_rating)
        previous_price=previous_price.append(previous_price)
        store_kink=store_kink.append(store_kink)
        discount=discount.append(discount)
        shipping_status=shipping_status.append(shipping_status)
        totalsold=totalsold.append(totalsold)
        rating=rating.append(rating)

data ={'product_name':product_name,
       'price':price,
       'item_rating':item_rating,
       'previous_price':previous_price,
       'store_link':store_link,
       'discount':discount,
       'shipping_status':shipping_status,
       'totalsold':totalsold,
       'rating':rating}
df = pd.DataFrame(data)


In [41]:

product_names = []
prices = []
star_ratings = []
previous_prices = []
store_links = []
discounts = []
shipping_statuses = []
totalsolds = []
ratings = []

totalpage = 10
for page in range(1, totalpage + 1):
    url = f"https://www.aliexpress.com/w/wholesale-phone.html?page={page}&g=y&SearchText=phone"
    driver.get(url)
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    totalphone = soup.find_all("div", class_="list--gallery--C2f2tvm search-item-card-wrapper-gallery")

    for phone in totalphone:
        product_name = phone.find("div", class_="multi--title--G7dOCj3").text.strip() if phone.find("div", class_="multi--title--G7dOCj3") else ''
        price = phone.find("div", class_="multi--price-sale--U-S0jtj").text.strip().replace('HUF', '').replace(',', '').replace('.', '') if phone.find("div", class_="multi--price-sale--U-S0jtj") else ''
        totalsold = phone.find('span', class_="multi--trade--Ktbl2jB").text.strip() if phone.find('span', class_="multi--trade--Ktbl2jB") else ''
        previous_price = phone.find('div', class_='multi--price-original--1zEQqOK').text.strip().replace('HUF', '').replace(',', '').replace('.', '') if phone.find('div', class_='multi--price-original--1zEQqOK') else ''
        shipping_status = phone.find("div", class_="multi--serviceContainer--3vRdzWN").text.strip() if phone.find("div", class_="multi--serviceContainer--3vRdzWN") else 'No shipping'
        
        try:
            rating = phone.find('div', class_='multi--starList--t9_CAj2').text.strip()
        except AttributeError:
            rating = ''
        
        try:
            store_link = phone.find("a", class_="cards--storeLink--XkKUQFS")['href']
        except (TypeError, KeyError):
            store_link = ''
        
        try:
            store_name = phone.find("a", class_="cards--storeLink--XkKUQFS").text.strip()
        except AttributeError:
            store_name = ''
        
        try:
            discount = phone.find("span", class_="multi--discount--3hksz5G").text.strip()
        except AttributeError:
            discount = ''
        try:
            star_rating = phone.find("div", class_="multi--starList--t9_CAj2").text.strip()
        except AttributeError:
            star_rating = ''
        
        # Append data to respective lists
        product_names.append(product_name)
        prices.append(price)
        star_ratings.append(store_name)
        previous_prices.append(previous_price)
        store_links.append(store_link)
        discounts.append(discount)
        shipping_statuses.append(shipping_status)
        totalsolds.append(totalsold)
        ratings.append(rating)

# Quit the browser
driver.quit()

# Create DataFrame
data = {
    'product_name': product_names,
    'price': prices,
    'star_rating': item_ratings,
    'previous_price': previous_prices,
    'store_link': store_links,
    'discount': discounts,
    'shipping_status': shipping_statuses,
    'totalsold': totalsolds,
    'rating': ratings
}

df = pd.DataFrame(data)
print(df.head())


                                        product_name    price  \
0  2023 New SERVO 15SE 3.0" Small Smartphone Dual...   129843   
1  SERVO Small Mobile Phone Bluetooth Dialer Magi...   416124   
2  MKTEL M2023 Feature Phone with 1.77inch Displa...   447674   
3  S25 Ultra New Smartphone 7.3 Inch Android 14 2...  2862016   
4  SERVO Mini Smart Phone Pure Android System 3.0...  1200138   

                      star_rating previous_price  \
0                     WFIVE Store        2375498   
1            SERVO Official Store         731395   
2  Desoon Technology Co., Limited         656535   
3            Shop1103411145 Store        3186102   
4            SERVO Official Store        3123278   

                              store_link discount  \
0  //www.aliexpress.com/store/1102821958     -45%   
1  //www.aliexpress.com/store/1102520433     -43%   
2  //www.aliexpress.com/store/1102625854     -31%   
3  //www.aliexpress.com/store/1103411145     -10%   
4  //www.aliexpress.com/store/1

data cleaning

chnage the name of the column

In [47]:
df.rename(columns={'star_rating':'store_name'})

product_name    price  \
0   2023 New SERVO 15SE 3.0" Small Smartphone Dual...   129843   
1   SERVO Small Mobile Phone Bluetooth Dialer Magi...   416124   
2   MKTEL M2023 Feature Phone with 1.77inch Displa...   447674   
3   S25 Ultra New Smartphone 7.3 Inch Android 14 2...  2862016   
4   SERVO Mini Smart Phone Pure Android System 3.0...  1200138   
..                                                ...      ...   
95  Original 6085 Mobile Cell Phone GSM 850/900/18...  1329712   
96  Blackberry 9720 Samoa 3G Mobile Cell Phone Ori...  1688999   
97  Original 5800 Xpress Music XpressMusic Mobile ...   122243   
98  Original Sony Xperia XZ F8331 F8332 SO-01J 4G ...  2716054   
99  Google Pixel 4a mobile phone Pixel4a Smartphon...  6335293   

                                  store_name previous_price  \
0                                WFIVE Store        2375498   
1                       SERVO Official Store         731395   
2             Desoon Technology Co., Limited         656535   
3                       Shop1103411145 Store        3186102   
4                       SERVO Official Store        3123278   
..                                       ...            ...   
95  BanMiDi Original Brand Phone Store Store                  
96  BanMiDi Original Brand Phone Store Store        2466231   
97  BanMiDi Original Brand Phone Store Store         179966   
98                       Stable Online Store                  
99                       Stable Online Store                  

                               store_link discount  \
0   //www.aliexpress.com/store/1102821958     -45%   
1   //www.aliexpress.com/store/1102520433     -43%   
2   //www.aliexpress.com/store/1102625854     -31%   
3   //www.aliexpress.com/store/1103411145     -10%   
4   //www.aliexpress.com/store/1102520433     -61%   
..                                    ...      ...   
95  //www.aliexpress.com/store/1102882129            
96  //www.aliexpress.com/store/1102882129     -31%   
97  //www.aliexpress.com/store/1102882129     -32%   
98  //www.aliexpress.com/store/1102634735            
99  //www.aliexpress.com/store/1102634735            

                                  shipping_status     totalsold rating  
0   Free shipping over HUF3,500 · 11-day delivery    3000+ sold         
1   Free shipping over HUF3,500 · 11-day delivery   5,000+ sold         
2   Free shipping over HUF3,500 · 11-day delivery  10,000+ sold         
3                                   Free shipping      278 sold         
4   Free shipping over HUF3,500 · 11-day delivery   2,000+ sold         
..                                            ...           ...    ...  
95                                  Free shipping        1 sold         
96                                  Free shipping       25 sold         
97                                  Free shipping       10 sold         
98                                  Free shipping        1 sold         
99                                  Free shipping        6 sold         

[100 rows x 9 columns]

In [57]:
df.store_link = df.store_link.str.lstrip('//')
df.store_link

0     www.aliexpress.com/store/1102821958
1     www.aliexpress.com/store/1102520433
2     www.aliexpress.com/store/1102625854
3     www.aliexpress.com/store/1103411145
4     www.aliexpress.com/store/1102520433
                     ...                 
95    www.aliexpress.com/store/1102882129
96    www.aliexpress.com/store/1102882129
97    www.aliexpress.com/store/1102882129
98    www.aliexpress.com/store/1102634735
99    www.aliexpress.com/store/1102634735
Name: store_link, Length: 100, dtype: object

In [63]:
df.discount=df.discount.str.rstrip('-').str.replace('-', '')
df.discount

0     45%
1     43%
2     31%
3     10%
4     61%
     ... 
95       
96    31%
97    32%
98       
99       
Name: discount, Length: 100, dtype: object

In [83]:
data={'shipping_fee':[x for x in df.shipping_status]}
data= pd.DataFrame(data)
data
data[['shipping_info', 'delivery_info']] = data['shipping_fee'].str.split('·', expand=True)
data

shipping_fee  \
0   Free shipping over HUF3,500 · 11-day delivery   
1   Free shipping over HUF3,500 · 11-day delivery   
2   Free shipping over HUF3,500 · 11-day delivery   
3                                   Free shipping   
4   Free shipping over HUF3,500 · 11-day delivery   
..                                            ...   
95                                  Free shipping   
96                                  Free shipping   
97                                  Free shipping   
98                                  Free shipping   
99                                  Free shipping   

                   shipping_info     delivery_info  
0   Free shipping over HUF3,500    11-day delivery  
1   Free shipping over HUF3,500    11-day delivery  
2   Free shipping over HUF3,500    11-day delivery  
3                  Free shipping              None  
4   Free shipping over HUF3,500    11-day delivery  
..                           ...               ...  
95                 Free shipping              None  
96                 Free shipping              None  
97                 Free shipping              None  
98                 Free shipping              None  
99                 Free shipping              None  

[100 rows x 3 columns]

In [85]:
data['free_shipping_over'] = data['shipping_info'].str.extract(r'over (HUF[\d,]+)')[0]
data

shipping_fee  \
0   Free shipping over HUF3,500 · 11-day delivery   
1   Free shipping over HUF3,500 · 11-day delivery   
2   Free shipping over HUF3,500 · 11-day delivery   
3                                   Free shipping   
4   Free shipping over HUF3,500 · 11-day delivery   
..                                            ...   
95                                  Free shipping   
96                                  Free shipping   
97                                  Free shipping   
98                                  Free shipping   
99                                  Free shipping   

                   shipping_info     delivery_info free_shipping_over  
0   Free shipping over HUF3,500    11-day delivery           HUF3,500  
1   Free shipping over HUF3,500    11-day delivery           HUF3,500  
2   Free shipping over HUF3,500    11-day delivery           HUF3,500  
3                  Free shipping              None                NaN  
4   Free shipping over HUF3,500    11-day delivery           HUF3,500  
..                           ...               ...                ...  
95                 Free shipping              None                NaN  
96                 Free shipping              None                NaN  
97                 Free shipping              None                NaN  
98                 Free shipping              None                NaN  
99                 Free shipping              None                NaN  

[100 rows x 4 columns]

In [87]:
data['delivery_days'] = data['delivery_info'].str.extract(r'(\d+)-day')[0]
data

shipping_fee  \
0   Free shipping over HUF3,500 · 11-day delivery   
1   Free shipping over HUF3,500 · 11-day delivery   
2   Free shipping over HUF3,500 · 11-day delivery   
3                                   Free shipping   
4   Free shipping over HUF3,500 · 11-day delivery   
..                                            ...   
95                                  Free shipping   
96                                  Free shipping   
97                                  Free shipping   
98                                  Free shipping   
99                                  Free shipping   

                   shipping_info     delivery_info free_shipping_over  \
0   Free shipping over HUF3,500    11-day delivery           HUF3,500   
1   Free shipping over HUF3,500    11-day delivery           HUF3,500   
2   Free shipping over HUF3,500    11-day delivery           HUF3,500   
3                  Free shipping              None                NaN   
4   Free shipping over HUF3,500    11-day delivery           HUF3,500   
..                           ...               ...                ...   
95                 Free shipping              None                NaN   
96                 Free shipping              None                NaN   
97                 Free shipping              None                NaN   
98                 Free shipping              None                NaN   
99                 Free shipping              None                NaN   

   delivery_days  
0             11  
1             11  
2             11  
3            NaN  
4             11  
..           ...  
95           NaN  
96           NaN  
97           NaN  
98           NaN  
99           NaN  

[100 rows x 5 columns]

In [91]:
data['free_shipping_over'].fillna('Free shipping', inplace=True)
data

shipping_fee  \
0   Free shipping over HUF3,500 · 11-day delivery   
1   Free shipping over HUF3,500 · 11-day delivery   
2   Free shipping over HUF3,500 · 11-day delivery   
3                                   Free shipping   
4   Free shipping over HUF3,500 · 11-day delivery   
..                                            ...   
95                                  Free shipping   
96                                  Free shipping   
97                                  Free shipping   
98                                  Free shipping   
99                                  Free shipping   

                   shipping_info     delivery_info free_shipping_over  \
0   Free shipping over HUF3,500    11-day delivery           HUF3,500   
1   Free shipping over HUF3,500    11-day delivery           HUF3,500   
2   Free shipping over HUF3,500    11-day delivery           HUF3,500   
3                  Free shipping              None      Free shipping   
4   Free shipping over HUF3,500    11-day delivery           HUF3,500   
..                           ...               ...                ...   
95                 Free shipping              None      Free shipping   
96                 Free shipping              None      Free shipping   
97                 Free shipping              None      Free shipping   
98                 Free shipping              None      Free shipping   
99                 Free shipping              None      Free shipping   

   delivery_days  
0             11  
1             11  
2             11  
3            NaN  
4             11  
..           ...  
95           NaN  
96           NaN  
97           NaN  
98           NaN  
99           NaN  

[100 rows x 5 columns]

In [94]:
data['free_shipping_over']=data['free_shipping_over'].str.replace('HUF','')

In [95]:
data

shipping_fee  \
0   Free shipping over HUF3,500 · 11-day delivery   
1   Free shipping over HUF3,500 · 11-day delivery   
2   Free shipping over HUF3,500 · 11-day delivery   
3                                   Free shipping   
4   Free shipping over HUF3,500 · 11-day delivery   
..                                            ...   
95                                  Free shipping   
96                                  Free shipping   
97                                  Free shipping   
98                                  Free shipping   
99                                  Free shipping   

                   shipping_info     delivery_info free_shipping_over  \
0   Free shipping over HUF3,500    11-day delivery              3,500   
1   Free shipping over HUF3,500    11-day delivery              3,500   
2   Free shipping over HUF3,500    11-day delivery              3,500   
3                  Free shipping              None      Free shipping   
4   Free shipping over HUF3,500    11-day delivery              3,500   
..                           ...               ...                ...   
95                 Free shipping              None      Free shipping   
96                 Free shipping              None      Free shipping   
97                 Free shipping              None      Free shipping   
98                 Free shipping              None      Free shipping   
99                 Free shipping              None      Free shipping   

   delivery_days  
0             11  
1             11  
2             11  
3            NaN  
4             11  
..           ...  
95           NaN  
96           NaN  
97           NaN  
98           NaN  
99           NaN  

[100 rows x 5 columns]

In [97]:
data = data.apply(lambda x: x.str.strip() if x.dtype == "object" else x)
data

shipping_fee  \
0   Free shipping over HUF3,500 · 11-day delivery   
1   Free shipping over HUF3,500 · 11-day delivery   
2   Free shipping over HUF3,500 · 11-day delivery   
3                                   Free shipping   
4   Free shipping over HUF3,500 · 11-day delivery   
..                                            ...   
95                                  Free shipping   
96                                  Free shipping   
97                                  Free shipping   
98                                  Free shipping   
99                                  Free shipping   

                  shipping_info    delivery_info free_shipping_over  \
0   Free shipping over HUF3,500  11-day delivery              3,500   
1   Free shipping over HUF3,500  11-day delivery              3,500   
2   Free shipping over HUF3,500  11-day delivery              3,500   
3                 Free shipping             None      Free shipping   
4   Free shipping over HUF3,500  11-day delivery              3,500   
..                          ...              ...                ...   
95                Free shipping             None      Free shipping   
96                Free shipping             None      Free shipping   
97                Free shipping             None      Free shipping   
98                Free shipping             None      Free shipping   
99                Free shipping             None      Free shipping   

   delivery_days  
0             11  
1             11  
2             11  
3            NaN  
4             11  
..           ...  
95           NaN  
96           NaN  
97           NaN  
98           NaN  
99           NaN  

[100 rows x 5 columns]

In [ ]:
data = data.drop(columns=['shipping_fee', 'delivery_info', 'shipping_info'])


In [101]:
data

free_shipping_over delivery_days
0               3,500            11
1               3,500            11
2               3,500            11
3       Free shipping           NaN
4               3,500            11
..                ...           ...
95      Free shipping           NaN
96      Free shipping           NaN
97      Free shipping           NaN
98      Free shipping           NaN
99      Free shipping           NaN

[100 rows x 2 columns]

In [104]:
data.columns

Index(['free_shipping_over', 'delivery_days'], dtype='object')

In [106]:
df[['free_shipping_over', 'delivery_days']] =data
df

product_name    price  \
0   2023 New SERVO 15SE 3.0" Small Smartphone Dual...   129843   
1   SERVO Small Mobile Phone Bluetooth Dialer Magi...   416124   
2   MKTEL M2023 Feature Phone with 1.77inch Displa...   447674   
3   S25 Ultra New Smartphone 7.3 Inch Android 14 2...  2862016   
4   SERVO Mini Smart Phone Pure Android System 3.0...  1200138   
..                                                ...      ...   
95  Original 6085 Mobile Cell Phone GSM 850/900/18...  1329712   
96  Blackberry 9720 Samoa 3G Mobile Cell Phone Ori...  1688999   
97  Original 5800 Xpress Music XpressMusic Mobile ...   122243   
98  Original Sony Xperia XZ F8331 F8332 SO-01J 4G ...  2716054   
99  Google Pixel 4a mobile phone Pixel4a Smartphon...  6335293   

                                 star_rating previous_price  \
0                                WFIVE Store        2375498   
1                       SERVO Official Store         731395   
2             Desoon Technology Co., Limited         656535   
3                       Shop1103411145 Store        3186102   
4                       SERVO Official Store        3123278   
..                                       ...            ...   
95  BanMiDi Original Brand Phone Store Store                  
96  BanMiDi Original Brand Phone Store Store        2466231   
97  BanMiDi Original Brand Phone Store Store         179966   
98                       Stable Online Store                  
99                       Stable Online Store                  

                             store_link discount  \
0   www.aliexpress.com/store/1102821958      45%   
1   www.aliexpress.com/store/1102520433      43%   
2   www.aliexpress.com/store/1102625854      31%   
3   www.aliexpress.com/store/1103411145      10%   
4   www.aliexpress.com/store/1102520433      61%   
..                                  ...      ...   
95  www.aliexpress.com/store/1102882129            
96  www.aliexpress.com/store/1102882129      31%   
97  www.aliexpress.com/store/1102882129      32%   
98  www.aliexpress.com/store/1102634735            
99  www.aliexpress.com/store/1102634735            

                                  shipping_status     totalsold rating  \
0   Free shipping over HUF3,500 · 11-day delivery    3000+ sold          
1   Free shipping over HUF3,500 · 11-day delivery   5,000+ sold          
2   Free shipping over HUF3,500 · 11-day delivery  10,000+ sold          
3                                   Free shipping      278 sold          
4   Free shipping over HUF3,500 · 11-day delivery   2,000+ sold          
..                                            ...           ...    ...   
95                                  Free shipping        1 sold          
96                                  Free shipping       25 sold          
97                                  Free shipping       10 sold          
98                                  Free shipping        1 sold          
99                                  Free shipping        6 sold          

   free_shipping_over delivery_days  
0               3,500            11  
1               3,500            11  
2               3,500            11  
3       Free shipping           NaN  
4               3,500            11  
..                ...           ...  
95      Free shipping           NaN  
96      Free shipping           NaN  
97      Free shipping           NaN  
98      Free shipping           NaN  
99      Free shipping           NaN  

[100 rows x 11 columns]

In [108]:
df.drop(columns=('shipping_status'), axis =1, inplace =True)

In [109]:
df

product_name    price  \
0   2023 New SERVO 15SE 3.0" Small Smartphone Dual...   129843   
1   SERVO Small Mobile Phone Bluetooth Dialer Magi...   416124   
2   MKTEL M2023 Feature Phone with 1.77inch Displa...   447674   
3   S25 Ultra New Smartphone 7.3 Inch Android 14 2...  2862016   
4   SERVO Mini Smart Phone Pure Android System 3.0...  1200138   
..                                                ...      ...   
95  Original 6085 Mobile Cell Phone GSM 850/900/18...  1329712   
96  Blackberry 9720 Samoa 3G Mobile Cell Phone Ori...  1688999   
97  Original 5800 Xpress Music XpressMusic Mobile ...   122243   
98  Original Sony Xperia XZ F8331 F8332 SO-01J 4G ...  2716054   
99  Google Pixel 4a mobile phone Pixel4a Smartphon...  6335293   

                                 star_rating previous_price  \
0                                WFIVE Store        2375498   
1                       SERVO Official Store         731395   
2             Desoon Technology Co., Limited         656535   
3                       Shop1103411145 Store        3186102   
4                       SERVO Official Store        3123278   
..                                       ...            ...   
95  BanMiDi Original Brand Phone Store Store                  
96  BanMiDi Original Brand Phone Store Store        2466231   
97  BanMiDi Original Brand Phone Store Store         179966   
98                       Stable Online Store                  
99                       Stable Online Store                  

                             store_link discount     totalsold rating  \
0   www.aliexpress.com/store/1102821958      45%    3000+ sold          
1   www.aliexpress.com/store/1102520433      43%   5,000+ sold          
2   www.aliexpress.com/store/1102625854      31%  10,000+ sold          
3   www.aliexpress.com/store/1103411145      10%      278 sold          
4   www.aliexpress.com/store/1102520433      61%   2,000+ sold          
..                                  ...      ...           ...    ...   
95  www.aliexpress.com/store/1102882129                 1 sold          
96  www.aliexpress.com/store/1102882129      31%       25 sold          
97  www.aliexpress.com/store/1102882129      32%       10 sold          
98  www.aliexpress.com/store/1102634735                 1 sold          
99  www.aliexpress.com/store/1102634735                 6 sold          

   free_shipping_over delivery_days  
0               3,500            11  
1               3,500            11  
2               3,500            11  
3       Free shipping           NaN  
4               3,500            11  
..                ...           ...  
95      Free shipping           NaN  
96      Free shipping           NaN  
97      Free shipping           NaN  
98      Free shipping           NaN  
99      Free shipping           NaN  

[100 rows x 10 columns]

In [116]:
df.totalsold=df.totalsold.str.replace('sold','')
df.totalsold

0       3000 
1      5,000 
2     10,000 
3        278 
4      2,000 
       ...   
95         1 
96        25 
97        10 
98         1 
99         6 
Name: totalsold, Length: 100, dtype: object

In [117]:
df.totalsold

0       3000 
1      5,000 
2     10,000 
3        278 
4      2,000 
       ...   
95         1 
96        25 
97        10 
98         1 
99         6 
Name: totalsold, Length: 100, dtype: object

In [118]:
df.head()

product_name    price  \
0  2023 New SERVO 15SE 3.0" Small Smartphone Dual...   129843   
1  SERVO Small Mobile Phone Bluetooth Dialer Magi...   416124   
2  MKTEL M2023 Feature Phone with 1.77inch Displa...   447674   
3  S25 Ultra New Smartphone 7.3 Inch Android 14 2...  2862016   
4  SERVO Mini Smart Phone Pure Android System 3.0...  1200138   

                      star_rating previous_price  \
0                     WFIVE Store        2375498   
1            SERVO Official Store         731395   
2  Desoon Technology Co., Limited         656535   
3            Shop1103411145 Store        3186102   
4            SERVO Official Store        3123278   

                            store_link discount totalsold rating  \
0  www.aliexpress.com/store/1102821958      45%     3000           
1  www.aliexpress.com/store/1102520433      43%    5,000           
2  www.aliexpress.com/store/1102625854      31%   10,000           
3  www.aliexpress.com/store/1103411145      10%      278           
4  www.aliexpress.com/store/1102520433      61%    2,000           

  free_shipping_over delivery_days  
0              3,500            11  
1              3,500            11  
2              3,500            11  
3      Free shipping           NaN  
4              3,500            11

In [123]:
df.head()

product_name    price  \
0  2023 New SERVO 15SE 3.0" Small Smartphone Dual...   129843   
1  SERVO Small Mobile Phone Bluetooth Dialer Magi...   416124   
2  MKTEL M2023 Feature Phone with 1.77inch Displa...   447674   
3  S25 Ultra New Smartphone 7.3 Inch Android 14 2...  2862016   
4  SERVO Mini Smart Phone Pure Android System 3.0...  1200138   

                      star_rating previous_price  \
0                     WFIVE Store        2375498   
1            SERVO Official Store         731395   
2  Desoon Technology Co., Limited         656535   
3            Shop1103411145 Store        3186102   
4            SERVO Official Store        3123278   

                            store_link discount totalsold free_shipping_over  \
0  www.aliexpress.com/store/1102821958      45%     3000               3,500   
1  www.aliexpress.com/store/1102520433      43%    5,000               3,500   
2  www.aliexpress.com/store/1102625854      31%   10,000               3,500   
3  www.aliexpress.com/store/1103411145      10%      278       Free shipping   
4  www.aliexpress.com/store/1102520433      61%    2,000               3,500   

  delivery_days  
0            11  
1            11  
2            11  
3           NaN  
4            11

check the data type of each columns

In [143]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   product_name        100 non-null    object
 1   price               100 non-null    int64 
 2   star_rating         100 non-null    object
 3   previous_price      100 non-null    object
 4   store_link          100 non-null    object
 5   discount            100 non-null    object
 6   totalsold           100 non-null    object
 7   free_shipping_over  100 non-null    object
 8   delivery_days       66 non-null     object
dtypes: int64(1), object(8)
memory usage: 7.2+ KB


In [163]:
df['previous_price'] = pd.to_numeric(df['previous_price'], errors='coerce')
df['totalsold'] = pd.to_numeric(df['totalsold'], errors='coerce').astype('float')
df['delivery_days'] = pd.to_numeric(df['delivery_days'], errors='coerce')




In [ ]:
df.free_shipping_over.fillna(0, inplace =True)
df.free_shipping_over



In [ ]:
df.delivery_days.fillna(0,inplace=True)
df.delivery_days


In [177]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   product_name        100 non-null    object 
 1   price               100 non-null    int64  
 2   star_rating         100 non-null    object 
 3   previous_price      84 non-null     float64
 4   store_link          100 non-null    object 
 5   discount            100 non-null    object 
 6   totalsold           73 non-null     float64
 7   free_shipping_over  100 non-null    float64
 8   delivery_days       100 non-null    float64
dtypes: float64(4), int64(1), object(4)
memory usage: 7.2+ KB


Q1: What is the average price of the phones listed

In [128]:
avg_price = df.price.astype(int).mean()
avg_price 

np.float64(1415235.04)

which phone have the higest price

In [187]:
df.groupby('product_name')['price'].sum().sort_values(ascending=False).head(4)

product_name
[World Premiere] CUBOT MAX 5, Android Smartphone 5G, Dimensity 8200, 6.95" 144Hz Large Screen, 24+256GB,Gaming Phone,Add to Cart    16200654
Original iPhone X 4G LTE CellPhone Unlocked 64GB/256GB ROM 5.8" A11 Bionic Chip Hexa Core FACE ID iphone x Mobile Cell Phone         7543856
FOSSiBOT F106 PRO Rugged Phone 6.58''FHD Display 12000mAh 30W Fast Charging 20MP Night Version With Loudly Speaker Smartphone        7445843
Original Samsung Galaxy A52s A528N 5G Mobile Phone NFC 6.5" 6GB+128GB 64MP+32MP Snapdragon 778G Octa-Core Android SmartPhone         6847117
Name: price, dtype: int64

phone with the minimum price

In [227]:
df.groupby('product_name')['price'].sum().sort_values(ascending=True).head(4)

product_name
Original 105 Dual Sim Good Quality 2G GSM 900/1800 Unlocked Mobile Cell Phone & No Hebrew Menu. No Network in North America           9145
SERVO Pocket S1 New Style Flip Mobile Phone 4 SIM Cards 2G GSM 2.4'' Screen Torch Auto Call Record Magic Voice Fold Phone Gifts      68682
SERVO Flip6 4 SIM Card Fold Mobile Phone GSM 2.4 Inch Display Auto Call Record  Speed Dial Magic Voice FM Lifted Cover Cellphone     75613
Original 6300 GSM 2G Mobile Cell Phone Russian & Arabic & Hebrew & English Keyboard Bluetooth Used CellPhone Made On 2007 Year      101147
Name: price, dtype: int64

which store have the higest price of phone 

In [189]:
df.groupby('star_rating')['price'].sum().sort_values(ascending=False).head(4)

star_rating
Stone's Store                               32904554
SERVO Official Store                        32699788
BanMiDi Original Brand Phone Store Store    18349318
Cubot Official Store                        16200654
Name: price, dtype: int64

In [190]:
df.groupby('star_rating')['price'].sum().sort_values(ascending=True).head(4)

star_rating
WFIVE Store                        129843
Desoon Technology Co., Limited     447674
SOYES Factory Choice Store        1408258
SuperGLX Store                    1441135
Name: price, dtype: int64

store with the higest shipping fee

In [191]:
df.groupby('star_rating')['free_shipping_over'].sum().sort_values(ascending=False).head(4)

star_rating
SERVO Official Store              122500.0
Stone's Store                      84000.0
SOYES Choice Store                 14000.0
Desoon Technology Co., Limited      3500.0
Name: free_shipping_over, dtype: float64

store with the minimum shipping fee

In [192]:
df.groupby('star_rating')['free_shipping_over'].sum().sort_values(ascending=True).head(4)

star_rating
BanMiDi Original Brand Phone Store Store    0.0
Cubot Official Store                        0.0
FOSSiBOT Official Store                     0.0
SUNSHINE Digital Store                      0.0
Name: free_shipping_over, dtype: float64

number of phone ship by each store 

In [194]:
df.groupby('star_rating')['free_shipping_over'].count().sort_values(ascending=False).head(7)

star_rating
SERVO Official Store                        38
Stone's Store                               29
BanMiDi Original Brand Phone Store Store    17
SOYES Choice Store                           4
Stable Online Store                          3
FOSSiBOT Official Store                      2
Cubot Official Store                         1
Name: free_shipping_over, dtype: int64

store with the minimum time of delivery

In [222]:
df.groupby(['star_rating','delivery_days'])['free_shipping_over'].count().sort_values( ascending =False)

star_rating                               delivery_days
SERVO Official Store                      11.0             28
Stone's Store                             11.0             23
BanMiDi Original Brand Phone Store Store  0.0              17
SERVO Official Store                      13.0              7
Stone's Store                             0.0               6
SOYES Choice Store                        11.0              4
SERVO Official Store                      0.0               3
Stable Online Store                       0.0               3
FOSSiBOT Official Store                   0.0               2
Cubot Official Store                      0.0               1
Desoon Technology Co., Limited            11.0              1
SOYES Factory Choice Store                11.0              1
SUNSHINE Digital Store                    0.0               1
Shop1103411145 Store                      0.0               1
SuperGLX Store                            11.0              1
WFIVE Store   

 Which store has the highest discount?



In [253]:
df1=df[['star_rating','discount']].value_counts().to_frame().sort_values(by ='discount',ascending=False).head(7)
df1

count
star_rating          discount       
Stone's Store        89%           1
                     73%           2
SOYES Choice Store   73%           1
SERVO Official Store 70%           2
Stone's Store        68%           1
                     67%           1
SERVO Official Store 66%           1

phone with the hiest discount

In [256]:
df2=df[['product_name','discount']].value_counts().to_frame().sort_values(by='discount',ascending=False).head(5)
df2

count
product_name                                                                                                                     discount       
SERVO Small Mobile Phone Bluetooth Dialer Magic voice Low Radiation Telephone Call Recording 2/3 SIM GSM Unlocked Mini Cellphone 89%           1
SERVO L8 Mini Laser Indicator Mobile Phone Bluetooth Dialing Magic Voice Blacklist Speed Dial 2 SIM 1.33" Small Cellphone Backup 73%           1
SERVO A50 Pro Flip Phone Dual SIM Card GSM Electric Torch Automatic Call Recording 2.4 Inch Screen Foldable Mobile Phones Type-C 73%           1
Q8 GSM 2G Rugged Mobile Phone Big Battery Double Flashlight 2.4inch HD Screen Dual Sim Large Button Cheap Cellphone For Elderly  73%           1
SERVO B25 Button Mini Mobile Phone 2 SIM Standby Bluetooth Dialing Surround Wear Blacklist Smallest Backup Cellphones for Sports 70%           1

average delivery time for the phone 

In [259]:
df['delivery_days'].mean()

np.float64(7.4)

store with the minimum delivery time 

In [265]:
df4=df[df['delivery_days']==0]
df4.groupby('star_rating')['delivery_days'].min()

star_rating
BanMiDi Original Brand Phone Store Store    0.0
Cubot Official Store                        0.0
FOSSiBOT Official Store                     0.0
SERVO Official Store                        0.0
SUNSHINE Digital Store                      0.0
Shop1103411145 Store                        0.0
Stable Online Store                         0.0
Stone's Store                               0.0
Name: delivery_days, dtype: float64

store with maximum delivery

In [266]:
df4=df[df['delivery_days']==11]
df4.groupby('star_rating')['delivery_days'].min()

star_rating
Desoon Technology Co., Limited    11.0
SERVO Official Store              11.0
SOYES Choice Store                11.0
SOYES Factory Choice Store        11.0
Stone's Store                     11.0
SuperGLX Store                    11.0
WFIVE Store                       11.0
Name: delivery_days, dtype: float64

From the analysis of the data, several key observations were made regarding the pricing, shipping, and discounts offered by various stores:

Highest and Lowest Priced Phones:

The store with the highest-priced phone is Stone's Store, with a phone priced at 32,904,554 HUF. This is followed closely by SERVO Official Store at 32,699,788 HUF.
On the other end of the spectrum, WFIVE Store and Desoon Technology Co., Limited offer the lowest-priced phones at 129,843 HUF and 447,674 HUF, respectively.
Shipping Fees:

BanMiDi Original Brand Phone Store and Cubot Official Store have been identified as the stores with the minimum shipping fees.
The maximum shipping fee observed is 3,500 HUF.
Delivery Time:

The average delivery time for phones to Hungary is 8 days.
However, it was noted that it takes 11 days for some deliveries.
Discounts:

Stone's Store offers a significant discount of 89%, while SOYES Choice Store provides a discount of 73% on all sales.
The product with the highest discount is the SERVO Small Mobile Phone Bluetooth Dialer Magic voice Low Radiation Telephone Call Recording 2/3 SIM GSM Unlocked Mini Cellphone.
Average Pricing:

The average price of phones across all stores is 1,415,235.04 HUF.e
4r48